File: contract/scripts/lib/flight-api.ts
/**
 * scripts/lib/flight-api.ts
 *
 * AviationStack REST API 래퍼
 * 실시간 항공편 지연 데이터를 가져옵니다.
 *
 * 무료 플랜: https://aviationstack.com (100 req/month, HTTP only)
 * 유료 플랜: HTTPS 지원, 더 많은 쿼리 허용
 *
 * 환경변수: AVIATIONSTACK_API_KEY
 */

export interface FlightDelayResult {
  /** 항공편 IATA 코드 (예: "KE017") */
  flightIata: string;
  /** 조회 기준 날짜 "YYYY-MM-DD" */
  flightDate: string;
  /** 출발 지연 시간 (분). 정시 출발이면 0. */
  departureDelay: number;
  /** 결항 여부 */
  cancelled: boolean;
  /** 원시 status 문자열 ("scheduled" | "active" | "landed" | "cancelled" | ...) */
  status: string;
  /** 예정 출발 시각 (ISO 8601) */
  scheduledDeparture: string;
  /** 예상 출발 시각 (null = 아직 미배정) */
  estimatedDeparture: string | null;
  /** 실제 출발 시각 (null = 미출발) */
  actualDeparture: string | null;
}

/**
 * AviationStack에서 특정 항공편의 지연 정보를 가져옵니다.
 *
 * @param apiKey  AVIATIONSTACK_API_KEY 환경변수 값
 * @param flightIata  항공편 IATA 코드 (예: "KE017")
 * @param flightDate  날짜 "YYYY-MM-DD". 생략 시 오늘 날짜 사용.
 * @returns FlightDelayResult, 또는 데이터 없으면 null
 */
export async function fetchFlightDelay(
  apiKey: string,
  flightIata: string,
  flightDate?: string
): Promise<FlightDelayResult | null> {
  const date = flightDate ?? new Date().toISOString().slice(0, 10);

  // 무료 플랜은 HTTP만 지원. flight_date는 유료 플랜 전용 파라미터이므로
  // URL에 포함하지 않고 응답 데이터에서 클라이언트 측 필터링으로 처리합니다.
  const url = new URL("http://api.aviationstack.com/v1/flights");
  url.searchParams.set("access_key", apiKey);
  url.searchParams.set("flight_iata", flightIata.toUpperCase());
  url.searchParams.set("limit", "10");

  const res = await fetch(url.toString());
  if (!res.ok) {
    throw new Error(`AviationStack HTTP error: ${res.status} ${res.statusText}`);
  }

  const body = (await res.json()) as any;

  // 에러 응답 처리
  if (body.error) {
    const msg = body.error.info ?? body.error.type ?? JSON.stringify(body.error);
    throw new Error(`AviationStack API error: ${msg}`);
  }

  if (!body.data || body.data.length === 0) {
    return null;
  }

  // 날짜 기준 클라이언트 필터링 (무료 플랜은 서버 side flight_date 미지원)
  const matched = body.data.filter(
    (d: any) => !d.flight_date || d.flight_date === date
  );
  if (matched.length === 0) {
    // 날짜 일치 없으면 가장 최근 항목 사용 (실시간 모니터링)
    console.warn(`[flight-api] ${date} 날짜 일치 없음 → 최근 데이터 사용`);
  }
  const f = (matched.length > 0 ? matched : body.data)[0];
  const rawDelay: number = f.departure?.delay ?? 0;

  return {
    flightIata: flightIata.toUpperCase(),
    flightDate: date,
    departureDelay: Math.max(0, rawDelay),
    cancelled: (f.flight_status ?? "") === "cancelled",
    status: f.flight_status ?? "unknown",
    scheduledDeparture: f.departure?.scheduled ?? "",
    estimatedDeparture: f.departure?.estimated ?? null,
    actualDeparture: f.departure?.actual ?? null,
  };
}

/** AviationStack API 키가 환경변수에 설정되어 있는지 확인하고 반환합니다. */
export function requireApiKey(): string {
  const key = process.env.AVIATIONSTACK_API_KEY;
  if (!key) {
    throw new Error(
      "AVIATIONSTACK_API_KEY 환경변수를 설정하세요.\n" +
        "  무료 가입: https://aviationstack.com\n" +
        "  실행 예시: AVIATIONSTACK_API_KEY=xxx yarn demo:oracle-resolve"
    );
  }
  return key;
}

File: backend/src/flight_api.rs
use anyhow::{bail, Context, Result};
use serde::Deserialize;

const BASE_URL: &str = "http://api.aviationstack.com/v1/flights";

#[derive(Debug, Clone)]
pub struct FlightDelayResult {
    pub flight_iata: String,
    pub flight_date: String,
    pub departure_delay: u16, // minutes (0 if not delayed)
    pub cancelled: bool,
    pub status: String,
}

#[derive(Deserialize)]
struct ApiResponse {
    data: Option<Vec<FlightData>>,
    error: Option<ApiError>,
}

#[derive(Deserialize)]
struct ApiError {
    message: String,
}

#[derive(Deserialize)]
struct FlightData {
    flight_date: Option<String>,
    flight_status: Option<String>,
    departure: Option<DepartureInfo>,
    flight: Option<FlightInfo>,
}

#[derive(Deserialize)]
struct DepartureInfo {
    delay: Option<i32>,
}

#[derive(Deserialize)]
struct FlightInfo {
    iata: Option<String>,
}

/// AviationStack API에서 항공편 지연 정보를 조회한다.
/// 무료 플랜: flight_date 서버 필터 미지원 → 클라이언트에서 필터링.
pub async fn fetch_flight_delay(
    api_key: &str,
    flight_iata: &str,
    flight_date: Option<&str>, // "YYYY-MM-DD"
) -> Result<Option<FlightDelayResult>> {
    if api_key.is_empty() {
        bail!("AVIATIONSTACK_API_KEY가 설정되지 않았습니다.");
    }

    let url = format!(
        "{BASE_URL}?access_key={api_key}&flight_iata={flight_iata}&limit=10"
    );

    let resp: ApiResponse = reqwest::get(&url)
        .await
        .context("AviationStack API 호출 실패")?
        .json()
        .await
        .context("AviationStack 응답 파싱 실패")?;

    if let Some(err) = resp.error {
        bail!("AviationStack API 오류: {}", err.message);
    }

    let data = match resp.data {
        Some(d) if !d.is_empty() => d,
        _ => return Ok(None),
    };

    // 날짜 필터 (클라이언트 측)
    let entry = if let Some(date) = flight_date {
        let matched: Vec<_> = data
            .iter()
            .filter(|d| d.flight_date.as_deref() == Some(date))
            .collect();
        if matched.is_empty() {
            tracing::warn!(
                "[flight_api] {flight_iata}: {date} 날짜 데이터 없음 → 최근 데이터 사용"
            );
            &data[0]
        } else {
            matched[0]
        }
    } else {
        &data[0]
    };

    let status = entry
        .flight_status
        .clone()
        .unwrap_or_else(|| "unknown".to_string());
    let cancelled = status == "cancelled";

    let raw_delay = entry
        .departure
        .as_ref()
        .and_then(|d| d.delay)
        .unwrap_or(0);
    // 음수(조기 출발)는 0으로, 10분 단위로 내림
    let delay_minutes = if raw_delay <= 0 {
        0u16
    } else {
        let d = raw_delay as u16;
        (d / 10) * 10
    };

    let iata = entry
        .flight
        .as_ref()
        .and_then(|f| f.iata.clone())
        .unwrap_or_else(|| flight_iata.to_string());

    let date = entry
        .flight_date
        .clone()
        .unwrap_or_else(|| "unknown".to_string());

    Ok(Some(FlightDelayResult {
        flight_iata: iata,
        flight_date: date,
        departure_delay: delay_minutes,
        cancelled,
        status,
    }))
}
